### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by fabian-hk**
Test environment with docker-compose and test cases defined as yaml files.

### CODE REVIEW

**Comment by fabian-hk** [comment]
I'm not sure if we should rename the test case equals_operator.yml to domains_and_ips.yml because it basically tests the module domains_and_ips.py. What do you think about this?

**Comment by fabian-hk** [comment]
I'm not sure exactly how the yaml files style guide works, but should we name the test cases with - instead of _?

**Comment by danielfett** [comment]
> I'm not sure if we should rename the test case equals_operator.yml to domains_and_ips.yml because it basically tests the module domains_and_ips.py. What do you think about this?

Okay for me!

**Comment by danielfett** [comment]
> I'm not sure exactly how the yaml files style guide works, but should we name the test cases with - instead of _?

I do not think that there is a consistent style. "-" is okay for me.

**Comment by danielfett** [comment]
I get this when trying to run run.py:
```
danielf@www:~/git/yesses$ ./run.py --help
Traceback (most recent call last):
  File "./run.py", line 6, in <module>
    from yesses import Runner, all_modules, Config
  File "/home/danielf/git/yesses/yesses/__init__.py", line 1, in <module>
    from .config import Config
  File "/home/danielf/git/yesses/yesses/config.py", line 3, in <module>
    from .findingslist import FindingsList
  File "/home/danielf/git/yesses/yesses/findingslist.py", line 2, in <module>
    from .state import State
  File "/home/danielf/git/yesses/yesses/state.py", line 3, in <module>
    import yesses.scan.tls_settings
  File "/home/danielf/git/yesses/yesses/scan/__init__.py", line 5, in <module>
    from .information_leakage import InformationLeakage
  File "/home/danielf/git/yesses/yesses/scan/information_leakage.py", line 5, in <module>
    from comment_parser import comment_parser
ModuleNotFoundError: No module named 'comment_parser'
```
Did you check in all files?

File: yesses/discover/error_paths.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            log.error("Could not open user agent list")
+            return
+
+        for origin in self.origins:
+            with utils.force_ip_connection(origin['domain'], origin['ip']):
+                parsed_url = utils.UrlParser(origin['url'])
+
+                with requests.Session() as req_sess:
+                    # get page with 404 not found error
+                    r = req_sess.get(f"{parsed_url.url_without_path}/fvwwvaoqgf/opdvsltqfnlcelh/ddsleo/glcgrfmr.odt",
```
Comment (danielfett) [inline yesses/discover/error_paths.py]: I propose to change the URL to:
/yesses-scanner-nonexisting-url/opdvsltqfnlcelh/ddsleo/glcgrfmr.html

File: yesses/discover/error_paths.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+from yesses.module import YModule, YExample
+from yesses import utils
+
+log = logging.getLogger('discover/error_paths')
+
+
+class ErrorPaths(YModule):
+    """
+    Collect error pages to scan them for information leakages.
```
Comment (danielfett) [inline yesses/discover/error_paths.py]: Please provide more information on what the module does.

File: yesses/discover/hidden_paths.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        self.dir_list = dir_list  # type: List[str]
+        self.pages_found = []  # type: List[utils.UrlParser]
+        self.dirs_found = []  # type: List[utils.UrlParser]
+
+
+class HiddenPaths(YModule):
+    """
+    This module takes urls and linked paths from the Linked Paths module.
+    It extracts potential folders from the linked paths and searches in
+    this folders with a wordlist for potential hidden files.
```
Comment (danielfett) [inline yesses/discover/hidden_paths.py]:
```suggestion
    these folders with a wordlist for potential hidden files.
```

File: yesses/discover/hidden_paths.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        self.linked_urls = [item['url'] for item in self.linked_paths]
+
+        filtered_origins = utils.filter_origins(self.origins)
+
+        for origin in filtered_origins.values():
+            with utils.force_ip_connection(origin['domain'], origin['ip']):
+                parsed_url = utils.UrlParser(origin['url'])
+
+                # check if the web server replies to a random path which should not exist with a 200 status
+                r = requests.get(f"{parsed_url.url_without_path}/fvwwvaoqgf/opdvsltqfnlcelh/ddsleo/glcgrfmr.odt",
```
Comment (danielfett) [inline yesses/discover/hidden_paths.py]: Similar URL to above.

File: yesses/utils.py [hunk 5]
```diff
@@ -21,5 +27,138 @@ def patched_create_connection(address, *args, **kwargs):
...
+    and second some web servers just redirect to another origin.
+    :param origins:
+    :return: origins without any duplication
+    """
+    filtered_origins = dict()
+    for origin in origins:
+        parsed_url = UrlParser(origin['url'])
+        with force_ip_connection(origin['domain'], origin['ip']):
+            r = requests.get(parsed_url.url_without_path)
+            forwareded_parsed_url = UrlParser(r.url)
```
Comment (danielfett) [inline yesses/utils.py]: forwar**e**ded → forwarded

File: yesses/utils.py [hunk 6]
```diff
@@ -21,5 +27,138 @@ def patched_create_connection(address, *args, **kwargs):
...
+
+        # retrieve the base domain (domain which was passed without the port and 'www.' prefix)
+        tmp = self.parsed.netloc.split(':')
+        self.base_domain = tmp[0]
+        split = self.base_domain.split('.')
+        if split[0] == "www":
+            self.base_domain = '.'.join(split[1:])
+
+        # the url with the protocol and port (if no port is specified use a standard port)
+        self.url_without_path = self.parsed.netloc
```
Comment (danielfett) [inline yesses/utils.py]: Wouldn't that be the origin?

Comment (fabian-hk) [reply yesses/utils.py]: I have just googled it and according to this [website](https://developer.mozilla.org/en-US/docs/Web/API/URL/origin), you are right. I will rename it.

**Comment by fabian-hk** [comment]
I think I have corrected all issues from the comments. I also moved the mini_config.yml file to docs/examples and renamed it to information_leakages_config.yml.

**Pull request merged**